File: epii-orm/src/map/WhereData.java
package epii.server.db.map;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import epii.server.db.Types;
import epii.server.db.tools.Tools;
 

public class WhereData extends MapData<WhereData> {
    protected List<String> expData = new ArrayList<>();
    public WhereData() {
        this.thisObj = this;
  }
    public WhereData putExp(String field, String exp) {
        expData.add(field + " = " + exp);
        return this;
    }

    public WhereData putLike(String field, String value) {
        expData.add(field + " like '" + value+"'");
        return this;
    }
    public WhereData putOp(String field,String op, String value) {
        expData.add(field + " "+op +" '" + value+"'");
        return this;
    }
    public WhereData putIn(String field, List condition, Types type){
        String condition_sql = "";
        if (type == Types.INT) {

            condition_sql = "(" + Tools.listToString(condition, ",") + ")";
        } else {

            condition_sql = "(\'" + Tools.listToString(condition, "\',\'") + "\')";
        }
        expData.add(field+" in "+condition_sql);

        return this;
    }
    public WhereData putNotIn(String field, List condition, Types type){
        String condition_sql = "";
        if (type == Types.INT) {

            condition_sql = "(" + Tools.listToString(condition, ",") + ")";
        } else {

            condition_sql = "(\'" + Tools.listToString(condition, "\',\'") + "\')";
        }
        expData.add(field+" not in  "+condition_sql);

        return this;
    }
    public WhereData putRaw(String rawWhere) {
        expData.add(rawWhere);
        return this;
    }

    public List<String> getExpData() {
        return expData;
    }

    public static WhereData fromMap(Map<String, Object> data) {
        WhereData fieldData = new WhereData();
        for (Map.Entry<String, Object> entry : data.entrySet()) {
            fieldData.put(entry.getKey(), entry.getValue());
        }
        return fieldData;
    }

    public static WhereData fromObject(Object data) {
        WhereData fieldData = new WhereData();
        for (Map.Entry<String, Object> entry : Tools.objectToMap(data).entrySet()) {
            fieldData.put(entry.getKey(), entry.getValue());
        }
        return fieldData;
    }

    public static WhereData make(String field, Object value) {
        WhereData data = new WhereData();
        data.put(field, value);
        return data;
    }

    public static WhereData make() {
        WhereData data = new WhereData();
        return data;
    }
}

File: epii-orm/src/map/WhereData.ts
import { BaseMap, PlainObject, WhereSymbol } from "../InterfaceTypes";

export class WhereData {
  mapData: BaseMap = new Map();
  expData: Array<string> = [];
  private initData(initData: PlainObject | BaseMap) {
    if (initData instanceof Map) {
      this.mapData = initData;
    } else if (typeof initData === "object") {
      for (const key in initData) {
        if (Object.prototype.hasOwnProperty.call(initData, key)) {
          this.mapData.set(key, initData[key]);
        }
      }
    }
  }
  constructor(initData: PlainObject | BaseMap | null = null) {
    if (initData === null) return;
    this.initData(initData);
  }
  put(field: string, value: string):WhereData{
    this.mapData.set(field,value);
    return this;
  }
  putExp(field: string, exp: string): WhereData {
    this.expData.push(field + " = " + exp);
    return this;
  }
  putLike(field: string, value: string): WhereData {
    this.expData.push(field + " like '" + value + "'");
    return this;
  }
  // 包含 id in (1,2,3)
  putIn(field: string, value: Array<string | Number>): WhereData {
    let condition_sql = "";

    if (typeof value[0] == "string") {
      for (let index = 0; index < value.length; index++) {
        condition_sql += "'" + value[index] + "'" + ",";
      }
    } else {
      for (let index = 0; index < value.length; index++) {
        condition_sql += value[index].toString() + ",";
      }
    }
    condition_sql = condition_sql.slice(0, condition_sql.length - 1);
    this.expData.push(field + " in " + "(" + condition_sql + ")");
    return this;
  }
  //不包含  id not in (1,2,3)
  putNotIn(field: string, value: Array<string | Number>): WhereData {
    let condition_sql = "";

    if (typeof value[0] == "string") {
      for (let index = 0; index < value.length; index++) {
        condition_sql += "'" + value[index] + "'" + ",";
      }
    } else {
      for (let index = 0; index < value.length; index++) {
        condition_sql += value[index].toString() + ",";
      }
    }
    condition_sql = condition_sql.slice(0, condition_sql.length - 1);
    this.expData.push(field + " not in " + "(" + condition_sql + ")");
    return this;
  }

  // 有符号的
  putSymbol(field: string, type: WhereSymbol, value: string): WhereData {
    this.expData.push(field + type + value);
    return this;
  }

  // 在范围之内
  putBetween(field: string, sValue: string, eValue: string): WhereData {
    let condition_sql = field + " between " + sValue + " and " + eValue;
    this.expData.push(condition_sql);
    return this;
  }

  // 不在范围之内
  putNotBetween(field: string, sValue: string, eValue: string): WhereData {
    let condition_sql = field + " not between " + sValue + " and " + eValue;
    this.expData.push(condition_sql);
    return this;
  }

  static make(keyOrData: string | null = null, value: string): WhereData {
    let out = new WhereData();
    if (keyOrData != null) {
      out.putExp(keyOrData, value);
    }
    return out;
  }
}
